cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/SourceControlMetadataMigrationService.java

```diff
@@ -0,0 +1,121 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Service for migrating source control metadata from app_spec table to source control table.
+ * Starting from 6.10.1 we have introduced a separate table for storing scm metadata. As this
+ * information was previously stored in app_spec table we need to run this migration
+ */
+public class SourceControlMetadataMigrationService extends AbstractScheduledService {
+
+  private ScheduledExecutorService executor;
+  private final long runInterval;
+  private final TransactionRunner transactionRunner;
+  private final Store store;
+
+  private static final Logger LOG = LoggerFactory.getLogger(SourceControlMetadataMigrationService.class);
+
+  @Inject
+  SourceControlMetadataMigrationService(CConfiguration cConf, Store store,
+      TransactionRunner transactionRunner) {
+    this.runInterval = cConf.getLong(
+        Constants.SourceControlManagement.METADATA_MIGRATION_INTERVAL_SECONDS);
+    this.transactionRunner = transactionRunner;
+    this.store = store;
+  }
+
+  @Override
+  protected final ScheduledExecutorService executor() {
+    executor = Executors.newSingleThreadScheduledExecutor(
+        Threads.createDaemonThreadFactory("scm-metadata-migration"));
+    return executor;
+  }
+
+  @Override
+  protected void runOneIteration() {
+    try {
+      migrateMetadata();
+      // stop if the migration is successfull otherwise retry after delay
+      this.stop();
+    } catch (Exception e) {
+      // Log and retry on exception
+      LOG.error("Failed to migrate source control metadata", e);
+    }
+  }
+
+  @Override
+  protected Scheduler scheduler() {
+    return Scheduler.newFixedRateSchedule(1, runInterval, TimeUnit.SECONDS);
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    if (executor != null) {
+      executor.shutdownNow();
+    }
+  }
+
+  private void migrateMetadata() {
+    LOG.info("Starting source control metadata migration at {}", Instant.now());
+    // Get all current metadata from app_spec table.
+    // This is to avoid running recursive transaction while trying to update scm table
+    Map<ApplicationReference, SourceControlMeta> metaMap = new HashMap<>();
+    TransactionRunners.run(transactionRunner, context -> {
+      metaMap.putAll(AppMetadataStore.create(context).getAllSourceControlMeta());
+    });
+
+    // For each metadata Update source control table with the metadata
+    // only if there is no current metadata in source control table
+    metaMap.forEach((appRef, scmMeta) -> {
+      TransactionRunners.run(transactionRunner, context -> {
+        NamespaceSourceControlMetadataStore metadataStore = NamespaceSourceControlMetadataStore.create(
+            context);
+        SourceControlMeta currentMeta = metadataStore.get(appRef);
+        if (currentMeta == null) {
+          LOG.debug("Migrating scm metadata {} for application {}", appRef, scmMeta);
+          if (scmMeta == null) {
```

Comment:
In which case would it be null? I see it's read from GSON. It must be a very special case for it to produce null.

Reply:
In ApplicationSepcification table we only added SCM meta when push/pulled happened. Hence this value can be null when the meta was never there in the old table. GSON.fromJson auto handles input null value